crates/holochain_types/src/chain.rs

```diff
@@ -46,6 +48,64 @@ pub struct ChainFilterIter {
     end: bool,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// A [`ChainFilter`] with the action sequences for the
+/// starting position and any `until` hashes.
+pub struct ChainFilterRange {
+    /// The filter for for this chain.
+    filter: ChainFilter,
+    /// The start of this range is the end of
+    /// the filter iterator.
+    /// The end of this range is the sequence of
+    /// the starting position hash.
+    range: RangeInclusive<u32>,
+    /// The start of the ranges type.
+    range_start_type: RangeStartType,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// The type of chain item that starts the range.
+enum RangeStartType {
+    /// The range starts from genesis.
+    Genesis,
+    /// The Range starts from an action where `take`
+    /// has reached zero.
+    Take,
+    /// The range starts from an action where an
+    /// `until` hash was found.
+    Until,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// Outcome of trying to find the action sequences in a filter.
+pub enum Sequences {
```

Comment:
Maybe this can be called `ChainFilterOutcome`?

Reply:
I'm actually not so sure about this name. It really is about finding the sequences for that hashes.

Reply:
It's just weird because these are "sequence numbers", and we sometimes call them "sequences" for short, but a sequence is a collection of numbers, so this makes it sounds like it's a collection of collections of numbers instead of just a collection of numbers. Guess it's more of a rough edge in the name of the original concept.